arch/arm/core/fault.c

```diff
@@ -660,64 +649,38 @@ void _Fault(const NANO_ESF *esf, u32_t exc_return)
 
 	if (exc_return & EXC_RETURN_RETURN_STACK_Secure) {
 		/* Exception entry occurred in Secure stack. */
-		FAULT_DUMP(reason, esf, fault);
+		reason = _FaultHandle(esf, fault);
 	} else {
-		/* Exception entry occurred in Non-Secure stack. Therefore, the
+		/* Exception entry occurred in Non-Secure stack. Therefore, 'esf'
+		 * holds the Secure stack information, however, the actual
 		 * exception stack frame is located in the Non-Secure stack.
 		 */
-		NANO_ESF *esf_ns;
+
+		/* Dump the Secure stack before handling the actual fault. */
+		SECURE_STACK_DUMP(esf);
+
+		/* Handle the actual fault.
+		 * Extract the correct stack frame from the Non-Secure state
+		 * and supply it to the fault handing function.
+		 */
 		if (exc_return & EXC_RETURN_MODE_THREAD) {
-			esf_ns = (NANO_ESF *)__TZ_get_PSP_NS();
+			esf = (NANO_ESF *)__TZ_get_PSP_NS();
 			if ((SCB->ICSR & SCB_ICSR_RETTOBASE_Msk) == 0) {
 				PR_EXC("RETTOBASE does not match EXC_RETURN\n");
 				_SysFatalErrorHandler(_NANO_ERR_HW_EXCEPTION, esf);
```

Comment:
nitpick, not related to this patch: Is it intended that these _SysFatalErrorHandler invocations don't have a return after them?
AFAICT it will be called again after it returns because there is another call site at the end of _Fault. I see five calls to SysFatalErrorHandler in this function, seems like you want a goto instead or something

Also: nothing in this C file should EVER be calling _SysFatalErrorHandler. It should be calling _NanoFatalErrorHandler instead.